lib/commanded/aggregates/multi.ex

```diff
@@ -88,57 +114,122 @@ defmodule Commanded.Aggregate.Multi do
         %AnEvent{item: item, total: aggregate.total + item}
       end)
 
+  ## Example with named steps
+
+      alias Commanded.Aggregate.Multi
+
+      aggregate
+      |> Multi.new()
+      |> Multi.execute(:start, fn aggregate ->
+        %AnEvent{item: nil, total: 0}
+      end)
+      |> Multi.reduce(:interesting_event, [1, 2, 3], fn aggregate, item ->
+        %AnEvent{item: item, total: aggregate.total + item}
+      end)
+      |> Multi.reduce([4, 5, 6], fn aggregate,
+                                    item,
+                                    %{
+                                      start: aggregate_state_after_start,
+                                      interesting_event: aggregate_state_after_interesting_event
+                                    } ->
+        %AnEvent{item: item, total: aggregate.total + item}
+      end)
+
   """
-  @spec reduce(Multi.t(), Enum.t(), function()) :: Multi.t()
-  def reduce(%Multi{} = multi, enumerable, execute_fun) when is_function(execute_fun, 2) do
+  @spec reduce(Multi.t(), atom(), Enum.t(), function()) :: Multi.t()
+  def reduce(multi, step_name \\ false, enumerable, execute_fun)
+
+  def reduce(%Multi{} = multi, step_name, enumerable, execute_fun)
+      when is_function(execute_fun, 2) do
+    Enum.reduce(enumerable, multi, fn item, %Multi{} = multi ->
+      execute(multi, step_name, &execute_fun.(&1, item))
+    end)
+  end
+
+  def reduce(%Multi{} = multi, step_name, enumerable, execute_fun)
+      when is_function(execute_fun, 3) do
     Enum.reduce(enumerable, multi, fn item, %Multi{} = multi ->
-      execute(multi, &execute_fun.(&1, item))
+      execute(multi, step_name, &execute_fun.(&1, item, &2))
     end)
   end
 
   @doc """
   Run the execute functions contained within the multi, returning the updated
-  aggregate state and all created events.
+  aggregate state, the aggregate state for each named step and all created events.
   """
   @spec run(Multi.t()) ::
-          {aggregate :: struct(), list(event :: struct())} | {:error, reason :: any()}
+          {aggregate :: struct(), steps :: %{atom() => struct()}, list(event :: struct())}
+          | {:error, reason :: any()}
```

Comment:
Removing the `steps` from the return tuple will allow the function contract to be unchanged and hence won't be a breaking change.

```elixir 
@spec run(Multi.t()) :: {aggregate :: struct(), list(event :: struct())} | {:error, reason :: any()}
```

It should be acceptable to do this since the returned `steps` value isn't actually being used outside of the tests.

Reply:
Makes sense, thanks for the feedback! I will prepare a new commit.

Reply:
Updates here: 28b089e8f580fdbfe81da2981bbcbbd2f6936620